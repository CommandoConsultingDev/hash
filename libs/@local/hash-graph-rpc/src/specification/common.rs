use std::future::{ready, Future};

use bytes::Bytes;
use error_stack::{Context, Report, ResultExt};
use hash_status::StatusCode;
use thiserror::Error;

use crate::harpc::{Decode, Encode, Stateful};

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[cfg_attr(feature = "specta", derive(specta::Type))]
#[cfg_attr(feature = "specta", specta(transparent))]
pub struct Error(hash_status::Status<serde_json::Value>);

impl<C> From<Report<C>> for Error
where
    C: Context,
{
    fn from(value: Report<C>) -> Self {
        let message = value.to_string();

        let status_code = value
            .request_ref::<StatusCode>()
            .next()
            .copied()
            .unwrap_or(StatusCode::Internal);

        let report = serde_json::to_value(&value).unwrap_or_else(|_| {
            tracing::error!("unable to convert error to json");
            serde_json::Value::Array(Vec::new())
        });

        let contents = match report {
            serde_json::Value::Array(reports) => reports,
            report => vec![report],
        };

        let status = hash_status::Status::new(status_code, Some(message), contents);

        Self(status)
    }
}

#[derive(Debug, Copy, Clone, Error)]
#[error("serde error")]
pub struct SerdeError;

pub struct JsonCodec;

impl Stateful for JsonCodec {
    type State = ();

    fn state(&self) -> Self::State {
        ()
    }
}

impl<T> Encode<T> for JsonCodec
where
    T: serde::Serialize,
{
    type Error = SerdeError;

    fn encode(
        &self,
        value: T,
    ) -> impl Future<Output = error_stack::Result<Bytes, Self::Error>> + Send {
        let bytes = serde_json::to_vec(&value)
            .change_context(SerdeError)
            .map(Bytes::from);

        ready(bytes)
    }
}

impl<T> Decode<T> for JsonCodec
where
    T: serde::de::DeserializeOwned + Send,
{
    type Error = SerdeError;

    fn decode(
        &self,
        bytes: Bytes,
    ) -> impl Future<Output = error_stack::Result<T, Self::Error>> + Send {
        let value = serde_json::from_slice(&bytes).change_context(SerdeError);

        ready(value)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JsonContext<S> {
    state: S,
}

impl<S> JsonContext<S> {
    #[must_use]
    pub const fn new(state: S) -> Self {
        Self { state }
    }
}

impl<S> Stateful for JsonContext<S>
where
    S: Send + Sync + Clone,
{
    type State = S;

    fn state(&self) -> Self::State {
        self.state.clone()
    }
}

impl<S, T> Encode<T> for JsonContext<S>
where
    S: Send + Sync + Clone,
    T: serde::Serialize,
{
    type Error = SerdeError;

    fn encode(
        &self,
        value: T,
    ) -> impl Future<Output = error_stack::Result<Bytes, Self::Error>> + Send {
        JsonCodec.encode(value)
    }
}

impl<S, T> Decode<T> for JsonContext<S>
where
    S: Send + Sync + Clone,
    T: serde::de::DeserializeOwned + Send,
{
    type Error = SerdeError;

    fn decode(
        &self,
        bytes: Bytes,
    ) -> impl Future<Output = error_stack::Result<T, Self::Error>> + Send {
        JsonCodec.decode(bytes)
    }
}

impl<S> crate::harpc::Context for JsonContext<S> where S: Clone + Send + Sync + 'static {}
