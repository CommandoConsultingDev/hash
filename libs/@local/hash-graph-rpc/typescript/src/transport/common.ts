import { Brand } from "effect";

export type TransportVersion = number & Brand.Brand<"TransportVersion">;
export const TransportVersion = Brand.refined<TransportVersion>(
  (value) => Number.isInteger(value) && value >= 0 && value <= 255,
  (value) =>
    Brand.error(`TransportVersion must be between 0 and 255, got ${value}`),
);

export type ServiceVersion = number & Brand.Brand<"ServiceVersion">;
export const ServiceVersion = Brand.refined<ServiceVersion>(
  (value) => Number.isInteger(value) && value >= 0 && value <= 255,
  (value) =>
    Brand.error(`ServiceVersion must be between 0 and 255, got ${value}`),
);

export type ServiceId = number & Brand.Brand<"ServiceId">;
export const ServiceId = Brand.refined<ServiceId>(
  (value) => Number.isSafeInteger(value) && value >= 0,
  (value) => Brand.error(`ServiceId must be a positive integer, got ${value}`),
);

export type ProcedureId = number & Brand.Brand<"ProcedureId">;
export const ProcedureId = Brand.refined<ProcedureId>(
  (value) => Number.isSafeInteger(value) && value >= 0,
  (value) =>
    Brand.error(`ProcedureId must be a positive integer, got ${value}`),
);

export type ActorId = Uint8Array & Brand.Brand<"ActorId">;
// validate that it is indeed a uuid, currently not done easily
export const ActorId = Brand.refined<ActorId>(
  (value) => value.length === 16,
  (value) =>
    Brand.error(`ActorId must be a 16 bytes UUID, got ${value.length}`),
);

export type PayloadSize = number & Brand.Brand<"Size">;
export const PayloadSize = Brand.refined<PayloadSize>(
  (value) => Number.isSafeInteger(value) && value >= 0,
  (value) => Brand.error(`Size must be a positive integer, got ${value}`),
);
