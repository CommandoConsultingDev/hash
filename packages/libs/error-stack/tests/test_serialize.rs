#![cfg(feature = "serde")]
// can be considered safe, because we only check the output, which in itself does not use **any**
// unsafe code.
#![cfg(not(miri))]
#![cfg_attr(all(nightly, feature = "std"), feature(error_generic_member_access))]
#![cfg_attr(nightly, feature(provide_any))]
// #![cfg_attr(nightly, feature(closure_lifetime_binder))]

mod common;

use common::snapshots::*;
use insta::assert_json_snapshot;

fn prepare(suffix: bool) -> impl Drop {
    let guard = snapshots::prepare(suffix, false, true);

    // backtraces are not consistent across platforms, and therefore super hard to test
    // for now we disable them
    std::env::set_var("RUST_LIB_BACKTRACE", "0");

    guard
}

/// This is the main test, to test all different parts at once,
/// and demonstrates that the rendering algorithm works at arbitrary depth.
#[test]
fn sources_nested() {
    let _guard = prepare(true);

    let report = create_sources_nested();

    assert_json_snapshot!(report);
}

#[cfg(all(
    rust_1_65,
    any(feature = "std", feature = "hooks"),
    feature = "spantrace",
))]
mod full {
    //! For reasoning about this specific module please refer to
    //! `test_debug.rs`
    use error_stack::{serde::HookContext, Report};

    use super::*;

    #[test]
    fn attachment() {
        let _guard = prepare(false);

        let report = create_report().attach_printable(PrintableA(2));

        assert_json_snapshot!(report);
    }

    #[test]
    fn context() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable(PrintableA(2))
            .change_context(ContextA(2));

        assert_json_snapshot!(report);
    }

    #[test]
    fn multiple_sources() {
        let _guard = prepare(false);

        let mut a = create_report().attach_printable(PrintableC(1));
        let b = create_report().attach_printable(PrintableC(2));

        a.extend_one(b);

        let a = a
            .attach_printable(PrintableC(3))
            .change_context(ContextA(2))
            .attach_printable(PrintableC(4));

        assert_json_snapshot!(a);
    }

    #[test]
    fn multiple_sources_at_root() {
        let _guard = prepare(false);

        let mut a = create_report().attach_printable(PrintableC(1));
        let b = create_report().attach_printable(PrintableC(2));

        a.extend_one(b);

        assert_json_snapshot!(a);
    }

    struct DoesNotImplementSerialize {
        a: String,
        b: Box<[u8]>,
    }

    #[derive(serde::Serialize)]
    struct ImplementSerialize<'a> {
        a: &'a str,
        b: &'a [u8],
    }

    #[derive(serde::Serialize)]
    struct ImplementSerializeOwned {
        a: String,
        b: Vec<u8>,
    }

    fn serialize<'a>(
        value: &'a DoesNotImplementSerialize,
        _: &mut HookContext<DoesNotImplementSerialize>,
    ) -> ImplementSerialize<'a> {
        ImplementSerialize {
            a: &value.a,
            b: &value.b,
        }
    }

    #[test]
    fn hook_custom() {
        let _guard = prepare(false);

        Report::install_custom_serde_hook(serialize);

        let report = create_report().attach(DoesNotImplementSerialize {
            a: "example".to_string(),
            b: Box::new([1, 2, 3]),
        });

        assert_json_snapshot!(report);
    }

    // #[test]
    // #[cfg(nightly)]
    // fn hook_custom_nightly() {
    //     let _guard = prepare(false);
    //
    //     Report::install_custom_serde_hook(
    //         for<'a, 'b> |value: &'a DoesNotImplementSerialize,
    //                      context: &'b mut HookContext<DoesNotImplementSerialize>|
    //                      -> ImplementSerialize<'a> {
    //             ImplementSerialize {
    //                 a: &value.a,
    //                 b: &value.b,
    //             }
    //         },
    //     );
    //
    //     let report = create_report().attach(DoesNotImplementSerialize {
    //         a: "example".to_string(),
    //         b: Box::new([1, 2, 3]),
    //     });
    //
    //     assert_json_snapshot!(report);
    // }

    #[test]
    fn hook_custom_owned() {
        let _guard = prepare(false);

        Report::install_custom_serde_hook(
            |value: &DoesNotImplementSerialize, _: &mut HookContext<DoesNotImplementSerialize>| {
                ImplementSerializeOwned {
                    a: value.a.clone(),
                    b: value.b.to_vec(),
                }
            },
        );

        let report = create_report().attach(DoesNotImplementSerialize {
            a: "example".to_string(),
            b: Box::new([1, 2, 3]),
        });

        assert_json_snapshot!(report);
    }

    // TODO: test auto/easy
    // TODO: test provider
    // TODO: test context serialize
}
