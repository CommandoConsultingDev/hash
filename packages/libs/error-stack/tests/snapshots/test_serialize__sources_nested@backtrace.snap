---
source: tests/test_serialize.rs
expression: report
---
[
  {
    "context": "context A",
    "attachments": [
      Location(
        file: "tests/common/snapshots.rs",
        line: 206,
        column: 10,
      ),
      "2",
      "1",
    ],
    "sources": [
      {
        "context": "context A",
        "attachments": [
          Location(
            file: "tests/common/snapshots.rs",
            line: 200,
            column: 10,
          ),
          "4",
          "3",
        ],
        "sources": [
          {
            "context": "root error",
            "attachments": [
              Location(
                file: "tests/common/mod.rs",
                line: 4,
                column: 5,
              ),
              "backtrace (1)",
              "6",
            ],
            "sources": [],
          },
        ],
      },
      {
        "context": "context A",
        "attachments": [
          Location(
            file: "tests/common/snapshots.rs",
            line: 195,
            column: 10,
          ),
          "5",
          "3",
        ],
        "sources": [
          {
            "context": "context A",
            "attachments": [
              Location(
                file: "tests/common/snapshots.rs",
                line: 193,
                column: 10,
              ),
              "7",
            ],
            "sources": [
              {
                "context": "context A",
                "attachments": [
                  Location(
                    file: "tests/common/snapshots.rs",
                    line: 184,
                    column: 34,
                  ),
                  "9",
                  "8",
                ],
                "sources": [
                  {
                    "context": "root error",
                    "attachments": [
                      Location(
                        file: "tests/common/mod.rs",
                        line: 4,
                        column: 5,
                      ),
                      "backtrace (1)",
                    ],
                    "sources": [],
                  },
                ],
              },
              {
                "context": "context A",
                "attachments": [
                  Location(
                    file: "tests/common/snapshots.rs",
                    line: 178,
                    column: 10,
                  ),
                  "13",
                  "10",
                  "16",
                  "9",
                  "8",
                ],
                "sources": [
                  {
                    "context": "root error",
                    "attachments": [
                      Location(
                        file: "tests/common/mod.rs",
                        line: 4,
                        column: 5,
                      ),
                      "backtrace (1)",
                    ],
                    "sources": [],
                  },
                ],
              },
              {
                "context": "context A",
                "attachments": [
                  Location(
                    file: "tests/common/snapshots.rs",
                    line: 164,
                    column: 10,
                  ),
                  "15",
                  "14",
                  "10",
                  "16",
                  "9",
                  "8",
                ],
                "sources": [
                  {
                    "context": "root error",
                    "attachments": [
                      Location(
                        file: "tests/common/mod.rs",
                        line: 4,
                        column: 5,
                      ),
                      "backtrace (1)",
                    ],
                    "sources": [],
                  },
                ],
              },
              {
                "context": "context A",
                "attachments": [
                  Location(
                    file: "tests/common/snapshots.rs",
                    line: 174,
                    column: 10,
                  ),
                  "11",
                  "9",
                  "8",
                ],
                "sources": [
                  {
                    "context": "root error",
                    "attachments": [
                      Location(
                        file: "tests/common/mod.rs",
                        line: 4,
                        column: 5,
                      ),
                      "backtrace (1)",
                    ],
                    "sources": [],
                  },
                ],
              },
              {
                "context": "context A",
                "attachments": [
                  Location(
                    file: "tests/common/snapshots.rs",
                    line: 170,
                    column: 10,
                  ),
                  "12",
                  "9",
                  "8",
                ],
                "sources": [
                  {
                    "context": "context A",
                    "attachments": [
                      Location(
                        file: "tests/common/snapshots.rs",
                        line: 169,
                        column: 10,
                      ),
                    ],
                    "sources": [
                      {
                        "context": "root error",
                        "attachments": [
                          Location(
                            file: "tests/common/mod.rs",
                            line: 4,
                            column: 5,
                          ),
                          "backtrace (1)",
                        ],
                        "sources": [],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]
